### ISSUE TITLE
questions about Farthest point sampling

### ISSUE THREAD

**Issue by pixar0407** [rusty1s/pytorch_cluster#102]
Hi, there.

Thank you for all the work you've done for graph community.
I have two questions which are basic though.

1) FPS is proposed firstly at pointnet++ paper?
at the source code of FPS(https://github.com/rusty1s/pytorch_cluster/blob/906497e66488901a7b67d24709f0551261c975b5/torch_cluster/fps.py#L19), you describe the FPS with pointnet++. So, is it mean that FPS is firstly used by pointnet++ OR FPS exists long before the pointnet++?

I guess FPS has long history itself. But i have to ask :)

2) Where is the code for FPS?
there is source code at here (https://github.com/rusty1s/pytorch_cluster/blob/906497e66488901a7b67d24709f0551261c975b5/torch_cluster/fps.py#L19). But i don't understand the line(49~70). Are the codes between line 49 ~ 79 actually all for the functionality of FPS? I think the lines are not enough. 
For example, there should be some content like this (https://github.com/yanx27/Pointnet_Pointnet2_pytorch/blob/bba1f6156371fbabf02bf4c47062dfde21a32b46/data_utils/ModelNetDataLoader.py#L25).

This is the end of my questions.
Thank you!

**Comment by rusty1s** [comment]
1. I guess FPS existed long before PointNet++, but AFAIK PointNet++ were the first to utilize it in a deep learning pipeline. The implementation is inspired by the one from PointNet++, that's why it is directly referenced.
2. You can find the implementation [here](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cpu/fps_cpu.cpp) (for CPU) and [here](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/fps_cuda.cu) (for GPU).

**Comment by pixar0407** [comment]
Thank you @rusty1s  :)

**Comment by pixar0407** [comment]
Hi again, @rusty1s 

After your reply, I have looked into the source codes ([gpu](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/fps_cuda.cu), [cpu](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cpu/fps_cpu.cpp)).

What i try to do is to compare the code performance between pytorch_cluster version fps and [this](https://github.com/yanx27/Pointnet_Pointnet2_pytorch/blob/bba1f6156371fbabf02bf4c47062dfde21a32b46/data_utils/ModelNetDataLoader.py#L25).

So, i have to understand the pytorch_cluster version fps.
before i go deep into the source codes that you have given to me earlier.
I have few questions.

1. how does [this code](https://github.com/rusty1s/pytorch_cluster/blob/906497e66488901a7b67d24709f0551261c975b5/torch_cluster/fps.py#L19) and the source codes ([gpu](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/fps_cuda.cu), [cpu](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cpu/fps_cpu.cpp)) are connected?

**Comment by rusty1s** [comment]
1. We call the C++/CUDA code via `torch.ops.torch_cluster.fps`, which gets linked to PyTorch [here](https://github.com/rusty1s/pytorch_cluster/blob/906497e66488901a7b67d24709f0551261c975b5/torch_cluster/__init__.py#L14-L15).
2. This is done [here](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/fps.cpp#L20-L29)
3. Yes.

**Comment by pixar0407** [comment]
it looks interesting. I will look into this. Thank you @rusty1s

**Comment by aluo-x** [comment]
Looking at the literature, I think the following was the first proposal for a similar algorithm:

- [Spectrally Optimal Sampling for Distribution Ray Tracing 1991, Don P. Mitchell](https://dl.acm.org/doi/10.1145/127719.122736)
    * this paper introduces the "Mitchell's best-candidate algorithm"
    * not to be confused with Scott A. Mitchell, who also publishes a lot of work on blue noise sampling
    * Don's 1987 paper "Generating antialiased images at low sampling densities" seems to be related, but I'm having trouble understanding this paper
    * Overall, the difference between Mitchell's best-candidate algorithm and the Farthest point sampling algorithm, seems to be an additional level of stochasticity, 

But the exact FPS algorithm seems to have been proposed in:
- [The Farthest Point Strategy for Progressive Image Sampling](https://ieeexplore.ieee.org/document/623193)
    * Seems to match the implementation in PointNet++ and this repo exactly
   
Currently the FPS step is too slow on large (1000k non-uniform points) clouds. I wonder if one of the methods below would be faster:

1. Use Mitchell best candidate (not sure how badly the non-uniform nature would hurt)
2. Sample points randomly in uniform within a range, select nearest neighbor in point cloud for each uniform point
3. Rejection sample. Sample points randomly in uniform within a range, select uniform points that are within epsilon of the point clouds, and only use those points' nearest neighbors in the point cloud

**Comment by pixar0407** [comment]
hi, @aluo-x 

You clearly pointed out all my concerns. I am focusing on efficiency (speed) of FPS algorithms on this repo and [pointnet++](https://github.com/erikwijmans/Pointnet2_PyTorch/blob/master/pointnet2_ops_lib/pointnet2_ops/_ext-src/src/sampling_gpu.cu). And I am also considering the possible alternatives(like Mitchell best candidate, as you mentioned before)
Thank you for the detailed answers.

By the way, can you give me the link of "Pointnet++" you had mentioned previously (i quoted below)?
because there are a lot of pointnet++ implementations. So, i just want to make sure... :) 

> The Farthest Point Strategy for Progressive Image Sampling
> Seems to match the implementation in PointNet++ and this repo exactly

**Comment by aluo-x** [comment]
PointConv [implementation here](https://github.com/DylanWusee/pointconv_pytorch/blob/4a01ee9e1742d4b837ab0ef44896be4e31c63378/utils/pointconv_util.py#L60)
    * seems to be copied from pointnet++ implementation below
 
[Pointnet++ implementation](https://github.com/yanx27/Pointnet_Pointnet2_pytorch/blob/bba1f6156371fbabf02bf4c47062dfde21a32b46/log/classification/pointnet2_ssg_wo_normals/pointnet2_utils.py#L63)
    * fairly straight forward

I should note that farthest point queries can be [accelerated by an R-tree](https://ieeexplore.ieee.org/abstract/document/7340928), also [discussed here](https://stackoverflow.com/a/59689579) and [here](https://stackoverflow.com/a/592397). I vaguely remember seeing a C# implementation a while back.

Another alternative which you can quickly implement using <5 lines of code, is to parallelize the farthest point sampling.

```
Given set C of candidate points, and we seek to find K farthest samples, and N cpu cores
1. Replicate C up to N times, such that each process has access to the same C candidate points.
2. On each core, start a new random seed. Start the FPS process, where each process finds K//N points using the standard FPS algorithm
3. Aggregate the results, optionally de-duplicate the points and repeat until K points are found.
```
This would obviously be non-optimal, but farthest point sampling is already non-optimal. I implemented it just now and visually the results look okay. Unfortunately for 1000k points, the parallel method is still too slow.

My problem:
1. Each cloud has up to 1000k very dense, non-uniform points. I want a reasonable uniform-ish point cloud of around 10k points.

My solution:
1. Initialize points uniformly inside unit sphere as query points (via rejection sampling). For each point, if the nearest neighbor in the candidate point-cloud is less than an epsilon (1e-2), then store the candidate point. Repeat and dedpulicate until we have a subset ~20k points. This step tries to get a reasonably uniform-ish point cloud from the initial non-uniform point cloud. Accelerated via `pynanoflann`.
2. Perform iterative farthest point sampling to select 10k points. This step tries to further clean up the uniform-ish point cloud to get an even more regular point cloud.


I'm running this offline, so the speed is reasonable. Unlike pure FPS sampling, which seems to be wildly unrealistic even when GPU accelerated.

I suspect Mitchell's best-candidate algorithm would also work fine. The method in the paper is to increase the "query" points at each iteration (constant times current number of selected points), but it seems like many implementations either use a fixed number or slightly different implementations. If query points == all points, then I think mitchell's algorithm is identical to farthest point sampling.

**Comment by pixar0407** [comment]
hi, @aluo-x 

Thank you for the answer and the further explanation. The content you have wrote seems quite interesting. I will look into them thoroughly.  After that, i will contact you later. Have good weekend.

**Comment by pixar0407** [comment]
hi, @aluo-x  and @rusty1s 

I had tried different versions of farthest point sampling codes from 3 versions.

1) [version1](https://github.com/yanx27/Pointnet_Pointnet2_pytorch/blob/bba1f6156371fbabf02bf4c47062dfde21a32b46/log/classification/pointnet2_ssg_wo_normals/pointnet2_utils.py#L63)
2) [version2 (this repository)](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/fps_cuda.cu)
3) [version3](https://github.com/Yochengliu/Relation-Shape-CNN/blob/6464eb8bb4efc686adec9da437112ef888e55684/utils/csrc/sampling_gpu.cu)

I compared the three versions of codes. And i found that all three have different performances.
I created random points [16,1024,3] ([B,N,C], B: batch size, N : number of points, C : point dimension) which means that batch size is 16, there are 1024 points in a batch, and each point is defined by 3 channel(i.e, 3).

And i sampled them from 1024 to 512. 
I calculate the time using 'time' module. And i only include the time of calculating the FPS. (excluding data preparation time and etc.)

So, the results are as follows,
1) 0.1574s
2) 0.00353s
3) 0.00005s

I was very surprised that the performance differences are so large. 
I have tried different configurations of [B,N,C], and the results are similar.
My question is why 2) and 3)'s performance difference is so large. I saw the both codes. They looks almost same to me, except for the 'best_dist' parts and the input dimension([B*N,C] vs [B,N,C]).

Any clue about that?
Thank you.

I will leave my code how i calculate the time.

2) version2

> import torch
> from time import time
> import numpy as np
> from torch_cluster import fps
> 
> 
> B = 16
> N = 4096
> C = 3
> to = 1024
> 
> clu_inputs = torch.rand((B*N,C),dtype=torch.float32).to(device='cuda')
> 
> clu_batch = torch.zeros(N,dtype=torch.long)
> for i in range(0,B-1):
>  tmp = torch.empty(N,dtype=torch.long).fill_(i+1)
>  clu_batch = torch.cat((clu_batch,tmp),0)
> clu_batch = clu_batch.to(device='cuda')
> 
> clu_ratio = torch.tensor(to/N,dtype=torch.float64,device='cuda')
> 
> print(clu_inputs.shape)
> print(clu_batch.shape)
> print(clu_ratio)
> 
> tmp_sum = 0
> for i in range(0,100):
>  clu_inputs = torch.rand((B*N,C),dtype=torch.float32).to(device='cuda')
>  ping = time()
>  fps_index = fps(clu_inputs,clu_batch,clu_ratio,False)
>  tmp_time = time()-ping
>  tmp_sum += tmp_time
>  print(tmp_time)
> 
> print('time : ', tmp_sum/100)


3) version3

> import sys
> sys.path.append('../utils')
> 
> import torch
> from time import time
> import pointnet2_utils
> 
> B = 16
> N = 4096
> C = 3
> to = 1024
> 
> xyz = torch.rand((B,N,C),dtype=torch.float32).to(device='cuda')
> print(xyz)
> ping = time()
> fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
> tmp_time = time()-ping
> print(tmp_time)
> 
> tmp_sum = 0
> for i in range(0,100):
>  xyz = torch.rand((B,N,C),dtype=torch.float32).to(device='cuda')
>  ping = time()
>  fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
>  tmp_time = time()-ping
>  tmp_sum += tmp_time
>  print(tmp_time)
> 
> print('done : ',tmp_sum/100)

**Comment by rusty1s** [comment]
Thanks, this is really interesting. I need to benchmark that by myself, but just some thoughts: Version3 is designed for 3-dimensional "features"/points and block size specific code, which allows CUDA to better optimize the given implementation.

In addition, for timing on CUDA, you should always make use of `torch.cuda.synchronize()` before calling `time()`.

**Comment by pixar0407** [comment]
Hi, @rusty1s .

Thank you for the comment. Oh, that's right! So, you mean that 'pytorch cluster' version spends '["dim"](https://github.com/rusty1s/pytorch_cluster/blob/master/csrc/cuda/fps_cuda.cu#L36)' more times than version 3, because version 3 considers dim is always [3](https://github.com/Yochengliu/Relation-Shape-CNN/blob/6464eb8bb4efc686adec9da437112ef888e55684/utils/csrc/sampling_gpu.cu#L119).

I wanna know if there is another reasons, if possible.
And i used time() on regular python file (xxx.py). then is it okay not to use torch.cuda.synchronize()?
But i will try using it and re-calculate the time :) thanks.

Let me share more experiment results.
batch is fixed to 16, and the colors are meaningless, just for visual comfort.
![image](https://user-images.githubusercontent.com/26530798/117658331-91f78400-b1d5-11eb-9fba-a17f38eeb022.png)

**Comment by pixar0407** [comment]
I added torch.cuda.synchronize() 2 time (i added code below).
This affect the total times in huge amount. Average time increases 100x. I don't think this is right.
But i will wait for your responce, because i want to accurately calculate the time.

> import sys
> sys.path.append('../utils')
> 
> import torch
> from time import time
> import pointnet2_utils
> 
> B = 16
> N = 8192
> C = 3
> to = 2048
> 
> # original inputs
> xyz = torch.rand((B,N,C),dtype=torch.float32).to(device='cuda')
> print(xyz)
> ping = time()
> fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
> tmp_time = time()-ping
> print(tmp_time)
> 
> tmp_sum = 0
> for i in range(0,100):
>  xyz = torch.rand((B,N,C),dtype=torch.float32).to(device='cuda')
>  **torch.cuda.synchronize()**
>  ping = time()
>  fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
>  **torch.cuda.synchronize()**
>  tmp_time = time()-ping
>  tmp_sum += tmp_time
>  print(tmp_time)

**Comment by rusty1s** [comment]
This looks correct to me.

**Comment by pixar0407** [comment]
> This looks correct to me.

"This" means new code that i added "torch.cuda.synchronize()" before every time() call?
wow.......... then i should re-calculate the whole experiments....

**Comment by rusty1s** [comment]
I'm sorry:(

CUDA is executed asynchronously: While the `fps` call may have been completed, CUDA execution might still not be finished. At worst, this mean you are currently measuring the runtime of calling a python function without really executing it.

**Comment by pixar0407** [comment]
hello, @rusty1s 

Thank you for the life-saving detail.
I have studied these [1](https://discuss.pytorch.org/t/why-time-time-in-python-is-inaccurte/94274/9?u=hohohihi), [2](https://discuss.pytorch.org/t/why-time-time-in-python-is-inaccurte/94274/9?u=hohohihi), [3](https://discuss.pytorch.org/t/how-to-measure-time-in-pytorch/26964), and re-calculate the duration.

![image](https://user-images.githubusercontent.com/26530798/117671434-c07c5b80-b1e3-11eb-84cc-818a2bf99a99.png)

And here is the code, again. (just in case that i might make mistake again.)

> xyz = torch.rand((B,N,C),dtype=torch.float32).to(device='cuda')
> fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
> fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
> fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
> 
> tmp_sum = 0
> for i in range(0,100):
>  xyz = torch.rand((B,N,C),dtype=torch.float32).to(device='cuda')
>  torch.cuda.synchronize()
>  ping = time.time()
>  fps_idx = pointnet2_utils.furthest_point_sample(xyz, to)
>  torch.cuda.synchronize()
>  tmp_time = time.time()-ping
>  tmp_sum += tmp_time
>  print(tmp_time)
> 
> print('done : ',tmp_sum/100)

**Comment by rusty1s** [comment]
Yes, this looks better. I looks like `torch_cluster.fps` is currently like 4 to 9 times slower than version 3. Is it possible for you to benchmark [`torch.ops.torch_cluster.fps`](https://github.com/rusty1s/pytorch_cluster/blob/master/torch_cluster/fps.py#L65-L70) directly (where `batch` is converted to `ptr` in a pre-processing step?

**Comment by aluo-x** [comment]
Slightly unrelated. May I suggest a possible improvement in the `fps` functioin interface in this library? 

The current design only allows a user to specify `ratio`, but it may be useful to also allow a user to specify an exact `num_pts`. It's not hard to slightly oversample using the current `fps` interface and then randomly subsample, but it is an unnecessary step.

**Comment by rusty1s** [comment]
Yes, we can allow that as well. Feel free to contribute if you find the time :)

**Comment by pixar0407** [comment]
hi, @aluo-x .

long time no see. 
recently i have looked closely into the link you had given to me. 

> I should note that farthest point queries can be accelerated by an R-tree, also discussed here and here. I vaguely remember seeing a C# implementation a while back.

I have a question. 
Do you mean "farthest point queries" is k-farthest neighbor  OR Farthest point sampling? 
Because k-farthest neighbor search is different from FPS. 
k-farthest neighbor search is opposite of k-nearest neighbor search.
But FPS is finding farthest for all the set points which is already chosen.
 
For example,
query = 0
pointset = {1,2,3,4,5,6,7,8,9,}

3-farthest neighbor = {7,8,9}
3-FPS = {0,9,5}, including query for simplicity.

So, i should reference this [link](https://stackoverflow.com/a/48926663/9336769).
This link mentions R-tree. But i think R-tree is used for finding the first point (like 9 in the example above). And R-tree does not help finding the next FPS point (like 5 in the example above).

I double R-tree would help accelerating FPS. 
What do you think about this?

**Comment by pixar0407** [comment]
hi, again. @aluo-x 

does this [paper](https://ieeexplore.ieee.org/document/623193)'s efficient algorithm is exact FPS or approximate FPS?

let me define 'exact FPS' and approx FPS.
'approx FPS' is what we use in pointnet++; we set first point in random (or just pick first one).
'exact FPS' is  real FPS; first pair should be selected using convex hull like in [here](https://stackoverflow.com/a/60955896/9336769)

Thank you.

**Comment by aluo-x** [comment]
> k-farthest neighbor search is opposite of k-nearest neighbor search.
> But FPS is finding farthest for all the set points which is already chosen.

I agree they are different, but the current FPS approach in pointnet++ uses is a greedy algorithm, so I suspect an R-tree would accelerate the (greedy) FPS search.



> Is this [paper](https://ieeexplore.ieee.org/document/623193)'s efficient algorithm is exact FPS or approximate FPS?

Not sure if it is really more efficient, but the paper seems to choose an initial point randomly, so it would be approx FPS. But does it really matter?
